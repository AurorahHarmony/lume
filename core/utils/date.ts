import { Temporal } from "../../deps/temporal.ts";

/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);
  const timestamp = parseInt(str) * 1000;

  if (timestamp) {
    return new Date(timestamp);
  }
}

/** Parse a string to Date */
export function parseDate(date: string): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string to a zoned datetime */
export function getZonedDateTime(
  date: string,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
