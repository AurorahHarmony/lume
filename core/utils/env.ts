export function env<T>(name: string): T | undefined {
  const value = Deno.env.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}
