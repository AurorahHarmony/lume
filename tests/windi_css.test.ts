import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import windiCSS from "../plugins/windi_css.ts";

Deno.test("Windi CSS plugin", async (t) => {
  const site = getSite({
    src: "windi_css",
  });

  site.use(windiCSS({
    config: {
      shortcuts: {
        "btn-green": "text-white bg-green-500 hover:bg-green-700",
      },
    },
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
