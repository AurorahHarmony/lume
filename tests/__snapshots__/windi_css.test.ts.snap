export const snapshot = {};

snapshot[`Windi CSS plugin 1`] = `
{
  formats: [
    {
      asset: true,
      engines: undefined,
      ext: ".windi.css",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.jsonc",
      pageLoader: [AsyncFunction: json],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".jsonc",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: undefined,
      ext: ".toml",
      pageLoader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
}
`;

snapshot[`Windi CSS plugin 2`] = `[]`;

snapshot[`Windi CSS plugin 3`] = `
[
  {
    content: "<!DOCTYPE html>
" +
      '<html lang="en"><head>
' +
      '  <meta charset="UTF-8">
' +
      '  <meta http-equiv="X-UA-Compatible" content="IE=edge">
' +
      '  <meta name="viewport" content="width=device-width, initial-scale=1.0">
' +
      "  <title>Windi demo</title>
" +
      '  <link rel="stylesheet" href="/windi.css">
' +
      "<style>*, ::before, ::after {
" +
      "  -webkit-box-sizing: border-box;
" +
      "  box-sizing: border-box;
" +
      "  border-width: 0;
" +
      "  border-style: solid;
" +
      "  border-color: #e5e7eb;
" +
      "}
" +
      "* {
" +
      "  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
" +
      "  --tw-ring-offset-width: 0px;
" +
      "  --tw-ring-offset-color: #fff;
" +
      "  --tw-ring-color: rgba(59, 130, 246, 0.5);
" +
      "  --tw-ring-offset-shadow: 0 0 #0000;
" +
      "  --tw-ring-shadow: 0 0 #0000;
" +
      "  --tw-shadow: 0 0 #0000;
" +
      "}
" +
      ":root {
" +
      "  -moz-tab-size: 4;
" +
      "  -o-tab-size: 4;
" +
      "  tab-size: 4;
" +
      "}
" +
      ":-moz-focusring {
" +
      "  outline: 1px dotted ButtonText;
" +
      "}
" +
      ":-moz-ui-invalid {
" +
      "  box-shadow: none;
" +
      "}
" +
      "::moz-focus-inner {
" +
      "  border-style: none;
" +
      "  padding: 0;
" +
      "}
" +
      "::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
" +
      "  height: auto;
" +
      "}
" +
      "::-webkit-search-decoration {
" +
      "  -webkit-appearance: none;
" +
      "}
" +
      "::-webkit-file-upload-button {
" +
      "  -webkit-appearance: button;
" +
      "  font: inherit;
" +
      "}
" +
      "[type='search'] {
" +
      "  -webkit-appearance: textfield;
" +
      "  outline-offset: -2px;
" +
      "}
" +
      "abbr[title] {
" +
      "  -webkit-text-decoration: underline dotted;
" +
      "  text-decoration: underline dotted;
" +
      "}
" +
      "a {
" +
      "  color: inherit;
" +
      "  text-decoration: inherit;
" +
      "}
" +
      "body {
" +
      "  margin: 0;
" +
      "  font-family: inherit;
" +
      "  line-height: inherit;
" +
      "}
" +
      "html {
" +
      "  -webkit-text-size-adjust: 100%;
" +
      '  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
' +
      "  line-height: 1.5;
" +
      "}
" +
      "p {
" +
      "  margin: 0;
" +
      "}
" +
      ".btn-green {
" +
      "  --tw-bg-opacity: 1;
" +
      "  background-color: rgba(16, 185, 129, var(--tw-bg-opacity));
" +
      "  --tw-text-opacity: 1;
" +
      "  color: rgba(255, 255, 255, var(--tw-text-opacity));
" +
      "}
" +
      ".btn-green:hover {
" +
      "  --tw-bg-opacity: 1;
" +
      "  background-color: rgba(4, 120, 87, var(--tw-bg-opacity));
" +
      "}
" +
      ".bg-purple-500 {
" +
      "  --tw-bg-opacity: 1;
" +
      "  background-color: rgba(139, 92, 246, var(--tw-bg-opacity));
" +
      "}</style></head>
" +
      "<body>
" +
      '  <div class="bg-purple-500">
' +
      "  <p>This is page 1</p>
" +
      "
" +
      '  <a href="page2.html" class="btn-green">Go to page 2</a>
' +
      "</div>
" +
      "
" +
      "</body></html>",
    data: {
      children: '<div class="bg-purple-500">
' +
        "  <p>This is page 1</p>
" +
        "
" +
        '  <a href="page2.html" class="btn-green">Go to page 2</a>
' +
        "</div>",
      content: '<div class="bg-purple-500">
' +
        "  <p>This is page 1</p>
" +
        "
" +
        '  <a href="page2.html" class="btn-green">Go to page 2</a>
' +
        "</div>",
      date: [],
      layout: "layout.njk",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/",
    },
    src: {
      asset: undefined,
      ext: ".njk",
      path: "/index",
      remote: undefined,
      slug: "index",
    },
  },
  {
    content: "<!DOCTYPE html>
" +
      '<html lang="en"><head>
' +
      '  <meta charset="UTF-8">
' +
      '  <meta http-equiv="X-UA-Compatible" content="IE=edge">
' +
      '  <meta name="viewport" content="width=device-width, initial-scale=1.0">
' +
      "  <title>Windi demo</title>
" +
      '  <link rel="stylesheet" href="/windi.css">
' +
      "<style>*, ::before, ::after {
" +
      "  -webkit-box-sizing: border-box;
" +
      "  box-sizing: border-box;
" +
      "  border-width: 0;
" +
      "  border-style: solid;
" +
      "  border-color: #e5e7eb;
" +
      "}
" +
      "* {
" +
      "  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
" +
      "  --tw-ring-offset-width: 0px;
" +
      "  --tw-ring-offset-color: #fff;
" +
      "  --tw-ring-color: rgba(59, 130, 246, 0.5);
" +
      "  --tw-ring-offset-shadow: 0 0 #0000;
" +
      "  --tw-ring-shadow: 0 0 #0000;
" +
      "  --tw-shadow: 0 0 #0000;
" +
      "}
" +
      ":root {
" +
      "  -moz-tab-size: 4;
" +
      "  -o-tab-size: 4;
" +
      "  tab-size: 4;
" +
      "}
" +
      ":-moz-focusring {
" +
      "  outline: 1px dotted ButtonText;
" +
      "}
" +
      ":-moz-ui-invalid {
" +
      "  box-shadow: none;
" +
      "}
" +
      "::moz-focus-inner {
" +
      "  border-style: none;
" +
      "  padding: 0;
" +
      "}
" +
      "::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
" +
      "  height: auto;
" +
      "}
" +
      "::-webkit-search-decoration {
" +
      "  -webkit-appearance: none;
" +
      "}
" +
      "::-webkit-file-upload-button {
" +
      "  -webkit-appearance: button;
" +
      "  font: inherit;
" +
      "}
" +
      "[type='search'] {
" +
      "  -webkit-appearance: textfield;
" +
      "  outline-offset: -2px;
" +
      "}
" +
      "abbr[title] {
" +
      "  -webkit-text-decoration: underline dotted;
" +
      "  text-decoration: underline dotted;
" +
      "}
" +
      "body {
" +
      "  margin: 0;
" +
      "  font-family: inherit;
" +
      "  line-height: inherit;
" +
      "}
" +
      "html {
" +
      "  -webkit-text-size-adjust: 100%;
" +
      '  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
' +
      "  line-height: 1.5;
" +
      "}
" +
      "p {
" +
      "  margin: 0;
" +
      "}
" +
      ".text-sky-100 {
" +
      "  --tw-text-opacity: 1;
" +
      "  color: rgba(224, 242, 254, var(--tw-text-opacity));
" +
      "}</style></head>
" +
      "<body>
" +
      '  <p class="text-sky-100">This is page 1</p>
' +
      "
" +
      "</body></html>",
    data: {
      children: '<p class="text-sky-100">This is page 1</p>',
      content: "content",
      date: [],
      layout: "layout.njk",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page/1.html",
    },
    src: {
      asset: undefined,
      ext: ".tmpl.js",
      path: "/multi[0]",
      remote: undefined,
      slug: "multi",
    },
  },
  {
    content: "<!DOCTYPE html>
" +
      '<html lang="en"><head>
' +
      '  <meta charset="UTF-8">
' +
      '  <meta http-equiv="X-UA-Compatible" content="IE=edge">
' +
      '  <meta name="viewport" content="width=device-width, initial-scale=1.0">
' +
      "  <title>Windi demo</title>
" +
      '  <link rel="stylesheet" href="/windi.css">
' +
      "<style>*, ::before, ::after {
" +
      "  -webkit-box-sizing: border-box;
" +
      "  box-sizing: border-box;
" +
      "  border-width: 0;
" +
      "  border-style: solid;
" +
      "  border-color: #e5e7eb;
" +
      "}
" +
      "* {
" +
      "  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
" +
      "  --tw-ring-offset-width: 0px;
" +
      "  --tw-ring-offset-color: #fff;
" +
      "  --tw-ring-color: rgba(59, 130, 246, 0.5);
" +
      "  --tw-ring-offset-shadow: 0 0 #0000;
" +
      "  --tw-ring-shadow: 0 0 #0000;
" +
      "  --tw-shadow: 0 0 #0000;
" +
      "}
" +
      ":root {
" +
      "  -moz-tab-size: 4;
" +
      "  -o-tab-size: 4;
" +
      "  tab-size: 4;
" +
      "}
" +
      ":-moz-focusring {
" +
      "  outline: 1px dotted ButtonText;
" +
      "}
" +
      ":-moz-ui-invalid {
" +
      "  box-shadow: none;
" +
      "}
" +
      "::moz-focus-inner {
" +
      "  border-style: none;
" +
      "  padding: 0;
" +
      "}
" +
      "::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
" +
      "  height: auto;
" +
      "}
" +
      "::-webkit-search-decoration {
" +
      "  -webkit-appearance: none;
" +
      "}
" +
      "::-webkit-file-upload-button {
" +
      "  -webkit-appearance: button;
" +
      "  font: inherit;
" +
      "}
" +
      "[type='search'] {
" +
      "  -webkit-appearance: textfield;
" +
      "  outline-offset: -2px;
" +
      "}
" +
      "abbr[title] {
" +
      "  -webkit-text-decoration: underline dotted;
" +
      "  text-decoration: underline dotted;
" +
      "}
" +
      "body {
" +
      "  margin: 0;
" +
      "  font-family: inherit;
" +
      "  line-height: inherit;
" +
      "}
" +
      "html {
" +
      "  -webkit-text-size-adjust: 100%;
" +
      '  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
' +
      "  line-height: 1.5;
" +
      "}
" +
      "p {
" +
      "  margin: 0;
" +
      "}
" +
      ".text-sky-200 {
" +
      "  --tw-text-opacity: 1;
" +
      "  color: rgba(186, 230, 253, var(--tw-text-opacity));
" +
      "}</style></head>
" +
      "<body>
" +
      '  <p class="text-sky-200">This is page 2</p>
' +
      "
" +
      "</body></html>",
    data: {
      children: '<p class="text-sky-200">This is page 2</p>',
      content: "content",
      date: [],
      layout: "layout.njk",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page/2.html",
    },
    src: {
      asset: undefined,
      ext: ".tmpl.js",
      path: "/multi[1]",
      remote: undefined,
      slug: "multi",
    },
  },
  {
    content: "<!DOCTYPE html>
" +
      '<html lang="en"><head>
' +
      '  <meta charset="UTF-8">
' +
      '  <meta http-equiv="X-UA-Compatible" content="IE=edge">
' +
      '  <meta name="viewport" content="width=device-width, initial-scale=1.0">
' +
      "  <title>Windi demo</title>
" +
      '  <link rel="stylesheet" href="/windi.css">
' +
      "<style>*, ::before, ::after {
" +
      "  -webkit-box-sizing: border-box;
" +
      "  box-sizing: border-box;
" +
      "  border-width: 0;
" +
      "  border-style: solid;
" +
      "  border-color: #e5e7eb;
" +
      "}
" +
      "* {
" +
      "  --tw-ring-inset: var(--tw-empty,/*!*/ /*!*/);
" +
      "  --tw-ring-offset-width: 0px;
" +
      "  --tw-ring-offset-color: #fff;
" +
      "  --tw-ring-color: rgba(59, 130, 246, 0.5);
" +
      "  --tw-ring-offset-shadow: 0 0 #0000;
" +
      "  --tw-ring-shadow: 0 0 #0000;
" +
      "  --tw-shadow: 0 0 #0000;
" +
      "}
" +
      ":root {
" +
      "  -moz-tab-size: 4;
" +
      "  -o-tab-size: 4;
" +
      "  tab-size: 4;
" +
      "}
" +
      ":-moz-focusring {
" +
      "  outline: 1px dotted ButtonText;
" +
      "}
" +
      ":-moz-ui-invalid {
" +
      "  box-shadow: none;
" +
      "}
" +
      "::moz-focus-inner {
" +
      "  border-style: none;
" +
      "  padding: 0;
" +
      "}
" +
      "::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
" +
      "  height: auto;
" +
      "}
" +
      "::-webkit-search-decoration {
" +
      "  -webkit-appearance: none;
" +
      "}
" +
      "::-webkit-file-upload-button {
" +
      "  -webkit-appearance: button;
" +
      "  font: inherit;
" +
      "}
" +
      "[type='search'] {
" +
      "  -webkit-appearance: textfield;
" +
      "  outline-offset: -2px;
" +
      "}
" +
      "abbr[title] {
" +
      "  -webkit-text-decoration: underline dotted;
" +
      "  text-decoration: underline dotted;
" +
      "}
" +
      "body {
" +
      "  margin: 0;
" +
      "  font-family: inherit;
" +
      "  line-height: inherit;
" +
      "}
" +
      "html {
" +
      "  -webkit-text-size-adjust: 100%;
" +
      '  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
' +
      "  line-height: 1.5;
" +
      "}
" +
      "p {
" +
      "  margin: 0;
" +
      "}
" +
      ".bg-purple-600 {
" +
      "  --tw-bg-opacity: 1;
" +
      "  background-color: rgba(124, 58, 237, var(--tw-bg-opacity));
" +
      "}</style></head>
" +
      "<body>
" +
      '  <div class="bg-purple-600">
' +
      "  <p>This is page 2</p>
" +
      "</div>
" +
      "
" +
      "</body></html>",
    data: {
      children: '<div class="bg-purple-600">
  <p>This is page 2</p>
</div>',
      content: '<div class="bg-purple-600">
  <p>This is page 2</p>
</div>',
      date: [],
      layout: "layout.njk",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page2/",
    },
    src: {
      asset: undefined,
      ext: ".njk",
      path: "/page2",
      remote: undefined,
      slug: "page2",
    },
  },
  {
    content: ".btn {
" +
      "  border-radius: 0.25rem;
" +
      "  font-weight: 700;
" +
      "  padding-top: 0.5rem;
" +
      "  padding-bottom: 0.5rem;
" +
      "  padding-left: 1rem;
" +
      "  padding-right: 1rem;
" +
      "}
" +
      ".btn-blue {
" +
      "  --tw-bg-opacity: 1;
" +
      "  background-color: rgba(59, 130, 246, var(--tw-bg-opacity));
" +
      "  --tw-text-opacity: 1;
" +
      "  color: rgba(255, 255, 255, var(--tw-text-opacity));
" +
      "  padding-top: 1rem;
" +
      "}
" +
      ".btn-blue:hover {
" +
      "  --tw-bg-opacity: 1;
" +
      "  background-color: rgba(29, 78, 216, var(--tw-bg-opacity));
" +
      "}",
    data: {
      content: ".btn {
" +
        "  @apply font-bold py-2 px-4 rounded;
" +
        "}
" +
        ".btn-blue {
" +
        "  @apply bg-blue-500 hover:bg-blue-700 text-white;
" +
        "  padding-top: 1rem;
" +
        "}",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/styles.css",
    },
    src: {
      asset: true,
      ext: ".windi.css",
      path: "/styles",
      remote: undefined,
      slug: "styles",
    },
  },
]
`;
